core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java

```diff
@@ -447,12 +527,71 @@ public Response createObjectOrUploadPart(@HeaderParam("Authorization") String au
       }
       AlluxioURI objectURI = new AlluxioURI(objectPath);
 
+      // Parse the TaggingData
+      TaggingData tagData = null;
+      if (tagging != null) { // PutObjectTagging
+        try {
+          tagData = new XmlMapper().readerFor(TaggingData.class).readValue(is);
+        } catch (IOException e) {
+          if (e.getCause() instanceof S3Exception) {
+            throw S3RestUtils.toObjectS3Exception((S3Exception) e.getCause(), objectPath);
+          }
+          throw new S3Exception(e, objectPath, S3ErrorCode.MALFORMED_XML);
+        }
+      }
+      if (taggingHeader != null) { // Parse the tagging header if it exists for PutObject
+        LOG.info("[czhu] tagging headers={}", taggingHeader);
+        List<TaggingData.TagObject> tagObjectList =
+            new ArrayList<TaggingData.TagObject>();
+        for (String tag : taggingHeader.split("&")) {
+          String[] entries = tag.split("=");
+          tagObjectList.add(new TaggingData.TagObject(entries[0], entries[1]));
+        }
+        try {
+          tagData = new TaggingData(new TaggingData.TagSet(tagObjectList));
+        } catch (IllegalArgumentException e) {
+          if (e.getCause() instanceof S3Exception) {
+            throw S3RestUtils.toObjectS3Exception((S3Exception) e.getCause(), objectPath);
+          }
+          throw S3RestUtils.toObjectS3Exception(e, objectPath);
+        }
+      }
+      LOG.info("[czhu] PutObjectTagging tagData={}", tagData);
+
+      // Populate the xattr Map with the metadata tags if provided
+      Map<String, ByteString> xattrMap = new HashMap<String, ByteString>();
+      if (tagData != null) {
+        try {
+          xattrMap.put(S3Constants.TAGGING_XATTR_KEY, TaggingData.serialize(tagData));
```

Comment:
I expect the usage pattern for tags to be write-once, read-many. So we serialize the entire `TaggingData` object and store that in `xAttr` as opposed to serializing the raw data in a `Map` or `List` and building a new object from that data in order to prioritize read efficiency over write efficiency.

- i.e: when tags are specified via the header `x-amz-tagging`, we have to wrap the tags into a `TaggingData` object before writing to `xAttr`. This means that all `GET` operations will be able to directly return the deserialized `xAttr` entry without any other data processing.